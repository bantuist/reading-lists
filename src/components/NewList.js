import React, { Component } from 'react';

// import './Lists/Lists.css';

class NewList extends Component {
  handleSubmit(event) {
    event.preventDefault();
    console.log(this);
    const name = this.input.value;

    this.props.onAddList(name);
    this.input.value = '';
  }
  render() {
    return (
      <div className="new-list-container">
        <form className="new-list-form" onSubmit={(event) => this.handleSubmit(event)}>
          <input
            className="new-list-input"
            ref={(input) => this.input = input}
            type="text"
            placeholder="Create a new list"
          />
        </form>
      </div>
    );
  }
}

export default NewList;
