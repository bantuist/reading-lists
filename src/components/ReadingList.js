import React, { Component } from 'react';
import './../App.css';

const fakeBooks = [
  {volumeInfo: {title: 'The Catcher in the Rye', authors: "J. D. Salinger", imageLinks: {thumbnail: 'http://books.google.com/books/content?id=PCDengEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api'}}}
];
class ReadingList extends Component {
  render() {
    const books = fakeBooks.map(book => {
      return (
        <li>
          <img 
            src={book.volumeInfo.imageLinks.thumbnail}
            alt="thumbnail"
          />
          {book.volumeInfo.title}
          {book.volumeInfo.authors}
          Reading List
        </li>
      );
    });
    return (
      <div className="col">
        <ul>
        {books}
        </ul>
      </div>
    );
  }
}

export default ReadingList;
