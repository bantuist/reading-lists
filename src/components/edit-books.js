import base from '../rebase';

export function addBook(newBook, currentList) {
  let bookId = newBook.id;
  let listId = currentList.id;
  // Make sure books isn't empty
  let books = {};
  if (currentList.books) {
    books = currentList.books;
  }
  // Check for duplicate book before adding
  if (books[`${bookId}`]) {
    console.log('Already added');
    return;
  }

  // Add new book
  books[`${bookId}`] = newBook;

  // Update books and bookCount
  currentList.books = books;
  currentList.bookCount++;

  base.post(`${listId}`, {
    data: currentList,
    context: this,
    then: () => {
      console.log('Posted to Firebase');
    }
  });
}

export function removeBook(bookId, currentList) {
  const { books } = currentList;
  const listId = currentList.id

  delete books[`${bookId}`];
  currentList.bookCount--;

  // vs. base.remove
  base.post(`${listId}`, {
    data: currentList,
    context: this,
    then: () => {
      console.log('Removed book');
    }
  });
}

export function updateBook(book, currentList) {
  const bookId = book.id;
  const listId = currentList.id;
  
  if (book.isRead) {
    currentList.books[`${bookId}`].isRead = false;
    currentList.doneCount--;
  } else {
    currentList.books[`${bookId}`].isRead = true;
    currentList.doneCount++;
  }

  base.update(`/${listId}`, {
    data: currentList,
    then(err) {
      if (!err) {
        console.log(`Updated book`);
      }
    }
  });

}