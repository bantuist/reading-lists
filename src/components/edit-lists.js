import base from '../rebase';
import uuidv1 from 'uuid/v1';

export function add(name) {
  const id = uuidv1();
  const createdAt = Date.now();
  const newList = { name, id, createdAt, bookCount: 0, doneCount: 0 };

  base.post(`/${id}`, {
    data: newList,
    context: this,
    then: () => {
      console.log(`Posted new list`);
    }
  });
}

export function remove(id) {
  base.remove(id).then(() => {
    console.log(`Removed list`);
  }).catch(error => {
    console.log('error', error);
  });
}

export function update(name, id) {
  base.update(`${id}`, {
    data: { name: name },
    then(err) {
      if (!err) {
        console.log(`Updated list`);
      }
    }
  });
}