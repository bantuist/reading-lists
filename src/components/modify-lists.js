import base from '../rebase';
import uuidv1 from 'uuid/v1';

export function addList(name, lists) {
  const id = uuidv1();
  const createdAt = Date.now();

  base.post(`/${id}`, {
    data: { name, id, createdAt, bookCount: 0, doneCount: 0 },
    context: this,
    then: () => {
      console.log(`Posted new list`);
    }
  });
}

export function removeList(id) {
  base.remove(id).then(() => {
    console.log(`Removed ${id}`);
  }).catch(error => {
    console.log('error', error);
  });
}

export function updateList(name, id) {
  base.update(`${id}`, {
    data: { name: name },
    then(err) {
      if (!err) {
        console.log(`Updated list`);
      }
    }
  });
}