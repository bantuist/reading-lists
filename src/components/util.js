// https://stackoverflow.com/questions/32911630/how-do-i-deal-with-localstorage-in-jest-tests
// class LocalStorageMock {
//   constructor() {
//     this.store = {};
//   }

//   clear() {
//     this.store = {};
//   }

//   getItem(key) {
//     return this.store[key];
//   }

//   setItem(key, value) {
//     this.store[key] = value;
//   }

//   removeItem(key) {
//     delete this.store[key];
//   }
// }

// console.log('Using LocalStorageMock');
// const localStorage = new LocalStorageMock();

export function getRequests() {
  let requests = JSON.parse(localStorage.getItem('requests'));
  // Track daily Google Books API calls
  if (!requests || requests.date !== new Date().toLocaleDateString()) {
    requests = {
      count: 0,
      date: new Date().toLocaleDateString()
    };
  } 
  return requests;
} 

export function setRequests(requests) {
  localStorage.setItem('requests', JSON.stringify(requests));
}

export function filterList(name, lists) {
  return Object.keys(lists).filter((key) => {
    return lists[key].name === name;
  });
}