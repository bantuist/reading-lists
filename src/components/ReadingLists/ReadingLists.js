import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import { add, remove, update } from '../edit-lists';
import { filterList } from '../util';
import base from '../../rebase';
import ReadingList from '../ReadingList/ReadingList';
import NewList from '../NewList/NewList';
import './ReadingLists.css';

class ReadingLists extends Component {
  constructor() {
    super();
    this.state = {
      lists: {},
      notifications: {
        editing: false,
        notify: false,
        message: ''
      }
    }

    this.addList = this.addList.bind(this);
    this.editList = this.editList.bind(this);
    this.removeList = this.removeList.bind(this);
    this.updateList = this.updateList.bind(this);
  }
  componentWillMount() {
    // Update local state props whenever base prop changes
    base.bindToState('/', {
      context: this,
      state: 'lists',
      asArray: false,
    });
  }
  // componentWillUnmount(){
  //   base.removeBinding(this.ref);
  // }
  addList(name) {
    const { lists } = this.state;
    const list = filterList(name, lists);

    // Check for duplicate list and blank name before adding new list
    if (list.length) {
      this.setState({ 
        notifications: { 
          notify: 'duplicate', 
          message: 'List already exists' 
        } 
      });
      setTimeout(() => {
        this.setState({ notifications: { notify: false } });
      }, 2000);
    } else if (!name) {
      this.setState({ notifications: { notify: 'blank' , message: 'List name can\'t be blank' } });
      setTimeout(() => {
        this.setState({ notifications: { notify: false } });
      }, 2000);
    } else {
      add(name);
    }
  }

  // Refactor
  updateList(event, id) {
    console.log('update');
    const name = event.target.value;

    if (!name) {
      this.setState({ 
        notifications: {
          notify: 'blank',
          message: 'List name can\t be blank'
        }
      }); 
      setTimeout(() => {
        this.setState({ notifications: { notify: false } });
      }, 2000);
    } else {
      update(name, id);
      this.setState({ notifications: { editing: false } });
    }
  }
  removeList(id) {
    remove(id);
  }
  editList(id, name) {
    this.setState({ notifications: { editing: id } });
    let listNode = ReactDOM
      .findDOMNode(this)
      .querySelector(`.list-item-edit[data-id='${id}']`
     );
    // Move cursor to end
    // const tempValue = name;
    listNode.value = '';
    // listNode.value = tempValue;
    // FIXME
    // TODO separate refs to focus input: ref={(input) => this.input = input} why doesn't this refer to a unique input in a list?
    // listNode.focus();
  }
  _sortLists(lists) {
    return Object.keys(this.state.lists).sort((a, b) => {
      return (lists[a]['createdAt'] < lists[b]['createdAt']);
    });
  }

  render() {
      let { lists, notifications } = this.state;
      let notifyClassName = 'notify';
      lists = this._sortLists(lists);
      
      // Map and render lists
      lists = lists.map(key => {
      const list = this.state.lists[key];
      const { id } = list;

      return (
        <ReadingList 
          key={id}
          list={list}
          notifications={notifications}
          onEditList={this.editList}
          onRemoveList={this.removeList}
          onUpdateList={this.updateList}
        />
      );
    });

    // New list notifications
    if (notifications.notify === 'duplicate' || notifications.notify === 'blank') {
      notifyClassName += ' duplicate';
    }

    return (
      <div className="wrapper">
        <ul className="list">
          <NewList onAddList={this.addList}/>
          <div className={notifyClassName}>{notifications.message}</div>
          {lists}
        </ul>
      </div>
    );
  }
}

export default ReadingLists;