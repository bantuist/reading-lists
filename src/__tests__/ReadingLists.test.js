import React from 'react';
import ReactDOM from 'react-dom';
import ReadingLists from '../components/ReadingLists/ReadingLists';
import { add, remove } from '../components/edit-lists';
import renderer from 'react-test-renderer';

test('ReadingLists should render ReadingLists', () => {
  const tree = renderer.create(<ReadingLists />).toJSON();
  expect(tree).toMatchSnapshot();
});

// test('addList adds a new list', () => {
  // const exampleList = ReadingLists.addList('Example List');
  // expect()
  // expect(exampleList).toEqual({
  //   name: expect.any(String),
  //   id: expect.any(String),
  //   createdAt: expect.any(Number),
  //   bookCount: expect.any(Number),
  //   doneCount: expect.any(Number),
  // });
// });

// test('removeList removes a list', () => {
//   const { lists } = this.state;
//   const exampleList = ReadingLists.addList('Example List');
//   expect(ReadingLists.removeList(exampleList.id)).toEqual({

//   });
// });

