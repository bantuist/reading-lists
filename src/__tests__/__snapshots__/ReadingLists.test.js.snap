// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadingLists should render ReadingLists 1`] = `
<div
  className="wrapper"
>
  <ul
    className="list"
  >
    <div
      className="new-list-container"
    >
      <form
        className="new-list-form"
        onSubmit={[Function]}
      >
        <input
          className="new-list-input"
          placeholder="Create a new list"
          type="text"
        />
      </form>
    </div>
    <div
      className="notify"
    >
      
    </div>
  </ul>
</div>
`;
